micro-rdk/src/common/ota.rs

```diff
@@ -334,71 +327,108 @@ impl<S: OtaMetadataStorage> OtaService<S> {
+    pub(crate) async fn update(&mut self) -> Result<bool, OtaError<S>> {
+        if !(self.needs_update()) {
+            return Ok(false);
+        }
+
+        let mut uri = self.parse_uri(&self.url)?;
+        let mut response: hyper::Response<hyper::body::Incoming>;
+
+        let conn = loop {
+            let (mut sender, conn) = loop {
+                let mut num_tries = 0;
+                num_tries += 1;
+                if num_tries == NUM_RETRY_CONN + 1 {
+                    return Err(OtaError::Other(
+                        "failed to establish connection".to_string(),
+                    ));
+                }
+                match self.connector.connect_to(&uri) {
+                    Ok(connection) => {
+                        match connection.await {
+                            Ok(io) => {
+                                match http2::Builder::new(self.exec.clone())
+                                    .max_frame_size(16_384) // lowest configurable
+                                    .timer(H2Timer)
+                                    .handshake(io)
+                                    .await
+                                {
+                                    Ok(pair) => break pair,
+                                    Err(e) => {
+                                        log::error!("failed to build http request: {}", e);
+                                    }
                                 }
                             }
-                        }
-                        Err(e) => {
-                            log::error!("failed to create tcp stream: {}", e);
+                            Err(e) => {
+                                log::error!("failed to create tcp stream: {}", e);
+                            }
                         }
                     }
-                }
-                Err(e) => {
-                    log::error!("failed to create http connection: {}", e);
-                }
-            };
+                    Err(e) => {
+                        log::error!("failed to create http connection: {}", e);
+                    }
+                };
 
-            log::debug!(
-                "retry establishing http connection to {} in {} seconds",
-                &uri,
-                CONN_RETRY_SECS
-            );
-            Timer::after(Duration::from_secs(CONN_RETRY_SECS)).await;
-        };
+                log::debug!(
+                    "retry establishing http connection to {} in {} seconds",
+                    &uri,
+                    CONN_RETRY_SECS
+                );
+                Timer::after(Duration::from_secs(CONN_RETRY_SECS)).await;
+            };
 
-        let conn = Box::new(self.exec.spawn(async move {
-            if let Err(err) = conn.await {
-                log::error!("connection failed: {:?}", err);
+            let conn = Box::new(self.exec.spawn(async move {
```

Comment:
From here to L430, all methods propagate errors in lieu of a `break` statement for the outer loop.